flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphGenerator.java

```diff
@@ -351,34 +356,67 @@ private void configureStreamGraph(final StreamGraph graph) {
         graph.setUserArtifacts(userArtifacts);
         graph.setTimeCharacteristic(timeCharacteristic);
         graph.setJobName(jobName);
-        graph.setJobType(shouldExecuteInBatchMode ? JobType.BATCH : JobType.STREAMING);
 
         if (shouldExecuteInBatchMode) {
-
-            if (checkpointConfig.isCheckpointingEnabled()) {
-                LOG.info(
-                        "Disabled Checkpointing. Checkpointing is not supported and not needed when executing jobs in BATCH mode.");
-                checkpointConfig.disableCheckpointing();
-            }
-
-            graph.setAllVerticesInSameSlotSharingGroupByDefault(false);
-            graph.setGlobalStreamExchangeMode(GlobalStreamExchangeMode.FORWARD_EDGES_PIPELINED);
+            configureStreamGraphBatch(graph);
             setDefaultBufferTimeout(-1);
-            setBatchStateBackendAndTimerService(graph);
         } else {
-            graph.setStateBackend(stateBackend);
-            graph.setChangelogStateBackendEnabled(changelogStateBackendEnabled);
-            graph.setCheckpointStorage(checkpointStorage);
-            graph.setSavepointDirectory(savepointDir);
+            configureStreamGraphStreaming(graph);
+        }
+    }
+
+    private void configureStreamGraphBatch(final StreamGraph graph) {
+        graph.setJobType(JobType.BATCH);
+
+        if (checkpointConfig.isCheckpointingEnabled()) {
+            LOG.info(
+                    "Disabled Checkpointing. Checkpointing is not supported and not needed when executing jobs in BATCH mode.");
+            checkpointConfig.disableCheckpointing();
+        }
+        setBatchStateBackendAndTimerService(graph);
+
+        graph.setGlobalStreamExchangeMode(deriveGlobalStreamExchangeModeBatch());
+        graph.setAllVerticesInSameSlotSharingGroupByDefault(false);
+    }
+
+    private void configureStreamGraphStreaming(final StreamGraph graph) {
+        graph.setJobType(JobType.STREAMING);
+
+        graph.setStateBackend(stateBackend);
+        graph.setChangelogStateBackendEnabled(changelogStateBackendEnabled);
+        graph.setCheckpointStorage(checkpointStorage);
+        graph.setSavepointDirectory(savepointDir);
+        graph.setGlobalStreamExchangeMode(deriveGlobalStreamExchangeModeStreaming());
+    }
+
+    private GlobalStreamExchangeMode deriveGlobalStreamExchangeModeBatch() {
+        final ShuffleMode shuffleMode = configuration.get(ExecutionOptions.SHUFFLE_MODE);
+        if (shuffleMode == ShuffleMode.ALL_EXCHANGES_PIPELINED) {
```

Comment:
Why not use switch-case here (and below)? Linting tools / IDEs would then also be able to show warnings/errors if the enum is not cased over exhaustively.